src/sqlfluff/core/parser/segments/base.py

```diff
@@ -743,34 +637,49 @@ def is_optional(cls) -> bool:
         return cls.optional
 
     @classmethod
-    def class_is_type(cls, *seg_type) -> bool:
+    def class_is_type(cls, *seg_type: str) -> bool:
         """Is this segment class (or its parent) of the given type."""
         # Use set intersection
         if cls._class_types.intersection(seg_type):
             return True
         return False
 
     @classmethod
-    def structural_simplify(cls, elem) -> Optional[dict]:
-        """Simplify the structure recursively so it serializes nicely in json/yaml."""
-        if len(elem) == 0:
-            return None
-        elif isinstance(elem, tuple):
-            # Does this look like an element?
-            if len(elem) == 2 and isinstance(elem[0], str):
-                # This looks like a single element, make a dict
-                elem = {elem[0]: cls.structural_simplify(elem[1])}
-            elif isinstance(elem[0], tuple):
-                # This looks like a list of elements.
-                keys = [e[0] for e in elem]
-                # Any duplicate elements?
-                if len(set(keys)) == len(keys):
-                    # No, we can use a mapping tuple
-                    elem = {e[0]: cls.structural_simplify(e[1]) for e in elem}
-                else:
-                    # Yes, this has to be a list :(
-                    elem = [cls.structural_simplify(e) for e in elem]
-        return elem
```

Comment:
This old method was _a mess_. I've basically totally re-written it to make it more manageable (and testable).